src/MOCKernel.cpp

```diff
@@ -0,0 +1,165 @@
+}
+
+
+/**
+ * @brief Constructor for the VolumeKernel assigns default values, calls
+ *        the MOCKernel constructor, and pulls refernces to FSR locks and FSR
+ *        volumes from the provided TrackGenerator.
+ * @param track_generator the TrackGenerator used to pull relevant tracking
+ *        data from
+ * @param row_num the row index into the temporary segments matrix
+ */
+VolumeKernel::VolumeKernel(TrackGenerator* track_generator, int row_num) :
+                           MOCKernel(track_generator, row_num) {
+  _FSR_locks = track_generator->getFSRLocks();
+
+  if (_FSR_locks == NULL)
+    log_printf(ERROR, "Unable to create a VolumeKernel without first creating "
+               "FSR locks");
+
+  _FSR_volumes = track_generator->getFSRVolumes();
+  _quadrature = track_generator->getQuadrature();
+  _weight = 0;
+}
+
+
+/**
+ * @brief Constructor for the CounterKernel assigns default values and calls
+ *        the MOCKernel constructor
+ * @param track_generator the TrackGenerator used to pull relevant tracking
+ *        data from
+ * @param row_num the row index into the temporary segments matrix
+ */
+CounterKernel::CounterKernel(TrackGenerator* track_generator, int row_num) :
+                           MOCKernel(track_generator, row_num) {}
+
+
+/**
+ * @brief Prepares an MOCKernel for a new Track
+ * @details Resets the segment count
+ * @param track The new Track the MOCKernel prepares to handle
+ */
+void MOCKernel::newTrack(Track* track) {
+  _count = 0;
+}
+
+
+/**
+ * @brief Prepares a VolumeKernel for a new Track
+ * @details Resets the segment count and updates the weight for the new Track
+ * @param track The new Track the MOCKernel prepares to handle
+ */
+void VolumeKernel::newTrack(Track* track) {
+
+  /* Compute the Track cross-sectional area */
+  int azim_index = track->getAzimAngleIndex();
+  _weight = _quadrature->getAzimSpacing(azim_index)
+      * _quadrature->getAzimWeight(azim_index);
+
+  /* Reset the count */
+  _count = 0;
+}
+
+
+/**
+ * @brief Destructor for MOCKernel
+ */
+MOCKernel::~MOCKernel() {};
+
+
+/*
+ * @brief Reads and returns the current count
+ * @details MOC kernels count how many times they are accessed. This value
+ *          returns the value of the counter (number of execute accesses)
+ *          since kernel creation or last call to newTrack.
+ * @return _count the counter value
+ */
+int MOCKernel::getCount() {
+  return _count;
+}
+
+
+/* @brief Resets the maximum optcal path length for a segment
+ * @details MOC kernels ensure that there are no segments with an optical path
+ *          length greater than the maximum optical path length by splitting
+ *          then when they get too large.
+ * @param the maximum optical path length for a segment
+ */
+void MOCKernel::setMaxOpticalLength(FP_PRECISION max_tau) {
+  _max_tau = max_tau;
+}
+
+
+/*
+ * @brief Adds segment contribution to the FSR volume
+ * @details The VolumeKernel execute function adds the product of the
```

Comment:
Add to the `@details` comment that this also cuts segments with lengths greater than the maximum permitted optical path length.